docs: update conflict detection documentation (#121)
osrd-project/osrd-website: content/docs/reference/design-docs/conflict-detection/index.en.md

This document is a work in progress
{{% /pageinfo %}}

Conflict detection is the process of looking for timetable conflicts.
A timetable conflict is any predictable condition which disrupts planned operations.

> Planned operations can be disrupted if a train slows down, is prevented from proceeding, or delayed.

Reviewer: ```suggestion
Planned operations can be disrupted if a train slows down, is prevented from proceeding, or is delayed.
```

Author: no need to repeat `is`. it parses as follows: `is (prevented from proceeding || delayed)`

Reviewer: The sentence structure and comma placement feel wrong in that case, I believe `X if A, B, or C` is almost always interpreted as `(A | B | C) => X`

What about "Planned operations can be disrupted if a train slows down, i.e. when it is delayed or prevented from proceeding"?

Reviewer 2: "Planned operations can be disrupted if a train is slowed down, prevented from proceeding, or delayed."

If you don't want to repeat the is.

Author: > I believe X if A, B, or C is almost always interpreted as (A | B | C) => X

That's the intent

> What about "Planned operations can be disrupted if a train slows down, i.e. when it is delayed or prevented from proceeding"?

It does not have the same meaning:
 - slowing down is for when the train is moving, slows down, but does not stop
 - delayed for when a departure cannot occur at the planned time
 - prevented from proceeding line about when a route is not set in time, and the train has to stop

Reviewer: What about this version?

> Planned operations can be disrupted if a train **either** slows down, is prevented from proceeding, or delayed.

I asked for Garance's input, apparently the `is` isn't necessary like I though it was, but it still a bit hard to parse the initial sentence right (which could cause misunderstandings). Adding an `either` makes it clearer.

Author: either does not quite work, as it's usually used to express an alternative between *two* options. How do you feel about @Khoyo's version?

Reviewer: I do like his version

Author: Alright, let's roll with it then

> Planned operations can be disrupted if a train slows down, is prevented from proceeding, or delayed.

Author: ```suggestion
Planned operations can be disrupted if a train is slowed down, prevented from proceeding, or delayed.
```

2) detect conflicting requirements

## Design constraints

The primary design goals are as follows:
 - enable threading new train paths into an existing timetable (see STDCM)
 - produce conflicts which can be linked back to a root cause
 - operate in way that can be visualized and interpreted
 - scale to real world timetables: millions of yearly trains, tens of thousands of daily trains

In addition to these goals, the following constraints apply:
 - it must be possible to thread new train paths into timetables with existing conflicts
 - it must not cause false-negatives: if no conflicts are detected, a multi-train simulation of the same timetable must not yield any slowdowns
 - it cannot rely on data we do not have
 - it has to enable later support of mobile block systems
 - it has to rely on existing signaling and interlocking simulation
 - it has to enable detecting conflicts regardless of the signaling system in use
 - it has to support transitions between signaling systems
 - it has to support conflicts between different signaling systems

## Conflict modeling

**Actors** are objects which cause resources to be used:

>  - **train paths** (or someone / something on the behalf of the train)

Author: train paths are scheduled trains. it's weird, I know

 - describe what the resource is needed for
 - detail how the resource is used, such as switch position, zone entry and exit

Resource requirements can turn out to be either **satisfied**
or **conflicting** with other requirements, depending on compatibility rules.

**Compatibility rules differ by requirement purpose and resource type**. For example:
 - spacing requirements are exclusive: simultaneous requirements for the same resource are conflicting
 - zone and switch requirements are shareable: simultaneous requirements are satisfied if the resource configuration is identical

**For conflict detection to work, resource requirements have to be at least as extensive as what's required to guarantee that a train path will not be disturbed.**

## Routing conflicts

### Context

For trains to proceed safely along their planned path:

 - switches have to be moved in the appropriate position
 - level crossings have to activate
 - risks of collision with other trains have to be mitigated

In practice, the path of trains is partitioned into routes, which when set, ensure a train can safely follow the route.

Routes have the following lifestyle:

>  - As a train approaches the start of one of its routes, it is **called** by an operator. If all resources required to safely use the route are available, switches and level crossings start to move. If a resources is not available, say because another train has reserved a section of track, this process is delayed until all conditions are met.

Reviewer: Nitpicky, feel free to ignore this. "say" sounds too familiar in this context in my opinion

```suggestion
 - As a train approaches the start of one of its routes, it is **called** by an operator. If all resources required to safely use the route are available, switches and level crossings start to move. If a resources is not available, e.g. because another train has reserved a section of track, this process is delayed until all conditions are met.
```

- For each zone in each route, simulate when it would be released, and thus not required anymore.

{{% pageinfo color="info" %}}
Route overlaps are not yet supported.
{{% /pageinfo %}}

### Requirement compatibility rules

Requirement compatibility is evaluated for all `RouteZoneRequirement`s, grouped by zone. Requirements A and B, ordered such that `A.set_deadline <= B.set_deadline`, are compatible if and only if either:

 - their active time span does not overlap, such that `A.release_time <= (B.set_deadline - activation_time)`, where the activation time is the delay required to reconfigure from `A.switches` to `B.switches`.
 - `(A.entry_det, A.exit_det, A.switches) == (B.entry_det, B.exit_det, B.switches)`

## Spacing conflicts

### Context

Even if interlocking mitigates some of the risks associated with operating trains, a major one is left out: head to tail collisions, caused by insufficient spacing.

This responsibility is handled by signaling, which conveys both interlocking and spacing constraints.

Signaling helps trains slow down until the end of their movement authority, which is either:

 - behind the tail of the next train

>  - at the end of the last route set for this train

Reviewer: Assuming the first option is for moving block, there's a missing case with the end of the last free block

Author: there must be a misunderstanding: the movement authority always ends either at the end of the last set route, or somewhere behind a train which also uses the same set route.
what do you mean by free block?

Reviewer: The misunderstanding was in the "somewhere behind a train" part. I understood this sentence as "the MA ends precisely at the tail of the next train".

But actually it is pretty clear, I just read it wrong.
